modules/imgproc/test/test_stackblur.cpp

```diff
@@ -0,0 +1,313 @@
+OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+
+#include "test_precomp.hpp"
+
+namespace opencv_test { namespace {
+
+template<typename T>
+void _stackblurRef(const Mat& src, Mat& dst, Size ksize)
+{
+    CV_Assert(!src.empty());
+    CV_Assert(ksize.width > 0 && ksize.height > 0 && ksize.height % 2 == 1 && ksize.width % 2 == 1);
+
+    dst.create(src.size(), src.type());
+    const int CN = src.channels();
+
+    int rowsImg = src.rows;
+    int colsImg = src.cols;
+    int wm = colsImg - 1;
+
+    int radiusW = ksize.width / 2;
+    int stackLenW = ksize.width;
+    const float mulW = 1.0f / (((float )radiusW + 1.0f) * ((float )radiusW + 1.0f));
+
+    // Horizontal direction
+    std::vector<T> stack(stackLenW * CN);
+    for (int row = 0; row < rowsImg; row++)
+    {
+        std::vector<float> sum(CN, 0);
+        std::vector<float> sumIn(CN, 0);
+        std::vector<float> sumOut(CN, 0);
+
+        const T* srcPtr = src.ptr<T>(row);
+
+        for (int i = 0; i <= radiusW; i++)
+        {
+            for (int ci = 0; ci < CN; ci++)
+            {
+                T tmp = *(srcPtr + ci);
+                stack[i * CN + ci] = tmp;
+                sum[ci] += tmp * (i + 1);
+                sumOut[ci] += tmp;
+            }
+        }
+
+        for (int i = 1; i <= radiusW; i++)
+        {
+            if (i <= wm) srcPtr += CN;
+            for(int ci = 0; ci < CN; ci++)
+            {
+                T tmp = *(srcPtr + ci);
+                stack[(i + radiusW) * CN + ci] = tmp;
+                sum[ci] += tmp * (radiusW + 1 - i);
+                sumIn[ci] += tmp;
+            }
+        }
+
+        int sp = radiusW;
+        int xp = radiusW ;
+        if (xp > wm) xp = wm;
+
+        T* dstPtr = dst.ptr<T>(row);
+        srcPtr = src.ptr<T>(row) + xp * CN;
+
+        int stackStart= 0;
+
+        for (int i = 0; i < colsImg; i++)
+        {
+            stackStart = sp + stackLenW - radiusW;
+
+            if (stackStart >= stackLenW) stackStart -= stackLenW;
+
+            for(int ci = 0; ci < CN; ci++)
+            {
+                *(dstPtr + ci) = cv::saturate_cast<T>(sum[ci] * mulW);
+                sum[ci] -= sumOut[ci];
+                sumOut[ci] -= stack[stackStart*CN + ci];
+            }
+
+            const T* srcNew = srcPtr;
+
+            if(xp < wm)
+                srcNew += CN;
+
+            for (int ci = 0; ci < CN; ci++)
+            {
+                stack[stackStart * CN + ci] = *(srcNew + ci);
+                sumIn[ci] += *(srcNew + ci);
+                sum[ci] += sumIn[ci];
+            }
+
+            int sp1 = sp + 1;
+            sp1 &= -(sp1 < stackLenW);
```

Comment:
It is a test code. So we don't need any kind of "optimization" obfuscation here.